browser/web_discovery/web_discovery_cta_util.cc

```diff
@@ -0,0 +1,136 @@
+#include "url/gurl.h"
+
+namespace {
+
+constexpr int kMaxDisplayCount = 5;
+constexpr char kWebDiscoveryCTAStateIdKey[] = "id";
+constexpr char kWebDiscoveryCTAStateCountKey[] = "count";
+constexpr char kWebDiscoveryCTAStateDismissedKey[] = "dismissed";
+constexpr char kWebDiscoveryCTAStateLastDisplayedKey[] = "last_displayed";
+
+bool IsBraveSearchDefault(TemplateURLService* template_service) {
+  DCHECK(template_service);
+
+  auto* template_url = template_service->GetDefaultSearchProvider();
+  if (!template_url)
+    return false;
+  return template_url->prepopulate_id() ==
+         TemplateURLPrepopulateData::PREPOPULATED_ENGINE_ID_BRAVE;
+}
+
+}  // namespace
+
+WebDiscoveryCTAState GetWebDiscoveryCTAState(PrefService* prefs,
+                                             const std::string& cta_id) {
+  WebDiscoveryCTAState state;
+  const auto& state_value = prefs->GetDict(kWebDiscoveryCTAState);
+  const auto* id = state_value.FindString(kWebDiscoveryCTAStateIdKey);
+  if (!id || *id != cta_id) {
+    // Give fresh state with new id.
+    state.id = cta_id;
+    return state;
+  }
+
+  state.id = *id;
+
+  const auto count = state_value.FindInt(kWebDiscoveryCTAStateCountKey);
+  const auto dismissed =
+      state_value.FindBool(kWebDiscoveryCTAStateDismissedKey);
+  const auto* last_displayed =
+      state_value.Find(kWebDiscoveryCTAStateLastDisplayedKey);
+  // If state is invalid, give fresh one.
+  if (!count || !dismissed || !last_displayed)
+    return state;
+
+  state.count = *count;
+  state.dismissed = *dismissed;
+  state.last_displayed = *base::ValueToTime(last_displayed);
+
+  return state;
+}
+
+std::string GetWebDiscoveryCurrentCTAId() {
+  if (!GetWebDiscoveryCTAIDForTesting().empty()) {
+    CHECK_IS_TEST();
+    return GetWebDiscoveryCTAIDForTesting();
+  }
+
+  // Update this when we want to start cta again.
+  // TODO(simonhong): Need to improve for updating current cta id.
+  // Maybe fetching new cta id remotely?
+  constexpr char kCurrentCTAId[] = "v1";
+  return kCurrentCTAId;
+}
+
+bool ShouldShowWebDiscoveryInfoBar(TemplateURLService* service,
+                                   PrefService* prefs,
+                                   const WebDiscoveryCTAState& state,
+                                   base::Clock* test_clock) {
+  if (prefs->GetBoolean(kWebDiscoveryEnabled))
+    return false;
+
+  if (!service || !IsBraveSearchDefault(service))
+    return false;
+
+  // Show when |state| is newly created one.
+  if (state.last_displayed.is_null() && state.count == 0)
+    return true;
+
+  // Don't show if user dismissed explicitely.
+  if (state.dismissed)
+    return false;
+
+  // Don't show same cta more than 5 times.
+  if (state.count >= kMaxDisplayCount)
+    return false;
+
+  const auto now = test_clock ? test_clock->Now() : base::Time::Now();
+
+  // Don't show if |last_displayed| is not valid.
+  // ex) last_displayed is newer than now.
+  if (now < state.last_displayed)
```

Comment:
Q. Is this possible?

Reply:
Maybe this will not be happened if user don't adjust pc's clock.
This is added just for safe. :)